### DISCUSSION TITLE
Trivy on EKS unable to scan any images

### DISCUSSION THREAD

**Discussion by mkm29** [aquasecurity/trivy#4237] [Bugs]
## Description

I am trying to scan all images deployed on my EKS cluster I am setting up for high security (will be deployed to classified IL5 environment). Kubernetes v1.23, all worker nodes run on Bottlerocket OS. 

## What did you expect to happen?

I expect images to be scanned and available in the VulnerabilityReports CRD

## What happened instead?

I was able to successfully install Falco to the cluster (uses containerd). However, when deploying the official Helm chart (0.6.0-rc3) the scan-vulnerability containers start and then immediately error out. I set this environment variable on the trivy-operator deployment:

```yaml
- name: CONTAINER_RUNTIME_ENDPOINT
  value: /run/containerd/containerd.sock
```

## Output of run with `-debug`:

```json
{
  "level": "error",
  "ts": 1668286646.865245,
  "logger": "reconciler.vulnerabilityreport",
  "msg": "Scan job container",
  "job": "trivy-system/scan-vulnerabilityreport-74f54b6cd",
  "container": "discovery",
  "status.reason": "Error",
  "status.message": "2022-11-12T20:57:13.674Z\t\u001b[31mFATAL\u001b[0m\timage scan error: scan error: unable to initialize a scanner: unable to initialize a docker scanner: 4 errors occurred:\n\t* unable to inspect the image (023620263533.dkr.ecr.us-gov-east-1.amazonaws.com/docker.io/istio/pilot:1.15.2): Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?\n\t* unable to initialize Podman client: no podman socket found: stat podman/podman.sock: no such file or directory\n\t* containerd socket not found: /run/containerd/containerd.sock\n\t* GET https://023620263533.dkr.ecr.us-gov-east-1.amazonaws.com/v2/docker.io/istio/pilot/manifests/1.15.2: unexpected status code 401 Unauthorized: Not Authorized\n\n\n\n",
  "stacktrace": "github.com/aquasecurity/trivy-operator/pkg/vulnerabilityreport.(*WorkloadController).processFailedScanJob\n\t/home/runner/work/trivy-operator/trivy-operator/pkg/vulnerabilityreport/controller.go:551\ngithub.com/aquasecurity/trivy-operator/pkg/vulnerabilityreport.(*WorkloadController).reconcileJobs.func1\n\t/home/runner/work/trivy-operator/trivy-operator/pkg/vulnerabilityreport/controller.go:376\nsigs.k8s.io/controller-runtime/pkg/reconcile.Func.Reconcile\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/reconcile/reconcile.go:102\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:121\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:320\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234"
}
```

I added the following security context:

```yaml
  seLinuxOptions:
    user: system_u
    role: system_r
    type: control_t
    level: s0-s0:c0.c1023
```

**Comment by mkm29** [comment]
I confirmed that bottlerocket uses containerd, as `/run/containerd/containerd.sock` is specified on my Falco deployment. Even when I mount this as volume onto the pod and set the `CONTAINER_RUNTIME_ENDPOINT` to this path I get the same error.

**Comment by 227235** [comment]
I have this problem too.  Any updates?

**Comment by knqyf263** [comment]
Is this relevant?
https://github.com/aquasecurity/trivy/issues/3059

**Comment by Talkapon** [comment]
I have the same problem with this image: quay.io/fluentd_elasticsearch/fluentd:v2.5.2. Nothing related to EKS. Simply running trivy on my MAC.

If I use docker pull quay.io/fluentd_elasticsearch/fluentd:v2.5.2 and then run trivy, everything works fine. It means that somehow trivy fails to retrieve an existing valid image.

Here is the error I get:
2022-12-13T12:50:07.027+0200	FATAL	image scan error:
    github.com/aquasecurity/trivy/pkg/commands/artifact.Run
        github.com/aquasecurity/trivy/pkg/commands/artifact/run.go:389
  - scan error:
    github.com/aquasecurity/trivy/pkg/commands/artifact.(*runner).scanArtifact
        github.com/aquasecurity/trivy/pkg/commands/artifact/run.go:249
  - unable to initialize a scanner:
    github.com/aquasecurity/trivy/pkg/commands/artifact.scan
        github.com/aquasecurity/trivy/pkg/commands/artifact/run.go:573
  - unable to initialize a docker scanner:
    github.com/aquasecurity/trivy/pkg/commands/artifact.imageStandaloneScanner
        github.com/aquasecurity/trivy/pkg/commands/artifact/scanner.go:22
  - 4 errors occurred:
	* unable to inspect the image (quay.io/fluentd_elasticsearch/fluentd:v2.5.2): Error: No such image: quay.io/fluentd_elasticsearch/fluentd:v2.5.2
	* unable to initialize Podman client: no podman socket found: stat podman/podman.sock: no such file or directory
	* containerd socket not found: /run/containerd/containerd.sock
	* unsupported MediaType: "application/vnd.docker.distribution.manifest.v1+prettyjws", see https://github.com/google/go-containerregistry/issues/377

**Comment by knqyf263** [comment]
@Talkapon Your error is something different. As the error message shows, the image is too old and not supported by google/containerregistry which Trivy uses for communicating with container registries.
https://github.com/google/go-containerregistry/issues/377

Docker CLI still supports those old media types. You can scan the image if you pull it like you said.

**Comment by Talkapon** [comment]
Thanks. Much appreciated

**Comment by knqyf263** [comment]
Please let us know if `CONTAINERD_NAMESPACE` helps.
https://github.com/aquasecurity/trivy/issues/3170#issuecomment-1315167685

**Comment by mkm29** [comment]
@knqyf263 in this case that variable did not help, it was the `seLinux` flags

**Comment by 227235** [comment]
Seems like it’s more complicated than that.
